indexing-service/src/main/java/io/druid/indexing/common/task/CompactionTask.java

```diff
@@ -0,0 +1,301 @@
+ */
+
+package io.druid.indexing.common.task;
+
+import com.fasterxml.jackson.annotation.JacksonInject;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.inject.Injector;
+import io.druid.data.input.impl.DimensionsSpec;
+import io.druid.indexing.common.TaskStatus;
+import io.druid.indexing.common.TaskToolbox;
+import io.druid.indexing.common.actions.SegmentListUsedAction;
+import io.druid.indexing.common.actions.TaskActionClient;
+import io.druid.indexing.common.task.IndexTask.IndexIOConfig;
+import io.druid.indexing.common.task.IndexTask.IndexIngestionSpec;
+import io.druid.indexing.common.task.IndexTask.IndexTuningConfig;
+import io.druid.indexing.firehose.IngestSegmentFirehoseFactory;
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.Pair;
+import io.druid.java.util.common.granularity.NoneGranularity;
+import io.druid.java.util.common.guava.Comparators;
+import io.druid.java.util.common.logger.Logger;
+import io.druid.query.aggregation.AggregatorFactory;
+import io.druid.segment.IndexIO;
+import io.druid.segment.QueryableIndex;
+import io.druid.segment.indexing.DataSchema;
+import io.druid.segment.indexing.granularity.ArbitraryGranularitySpec;
+import io.druid.segment.indexing.granularity.GranularitySpec;
+import io.druid.segment.loading.SegmentLoadingException;
+import io.druid.timeline.DataSegment;
+import io.druid.timeline.TimelineObjectHolder;
+import io.druid.timeline.VersionedIntervalTimeline;
+import io.druid.timeline.partition.PartitionChunk;
+import io.druid.timeline.partition.PartitionHolder;
+import org.joda.time.Interval;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+public class CompactionTask extends AbstractTask
+{
+  private static final Logger log = new Logger(CompactionTask.class);
+  private static final String TYPE = "compact";
+
+  private final Interval interval;
+  private final DimensionsSpec dimensionsSpec;
+  private final IndexTuningConfig tuningConfig;
+  private final Injector injector;
+  private final IndexIO indexIO;
+  private final ObjectMapper jsonMapper;
+
+  @JsonIgnore
+  private IndexTask indexTaskSpec;
+
+  @JsonCreator
+  public CompactionTask(
+      @JsonProperty("id") final String id,
+      @JsonProperty("resource") final TaskResource taskResource,
+      @JsonProperty("dataSource") final String dataSource,
+      @JsonProperty("interval") final Interval interval,
+      @JsonProperty("dimensionsSpec") final DimensionsSpec dimensionsSpec,
```

Comment:
Rather than asking the user to include this, it could be determined by looking at the dimensions and column types from each segment's StorageAdapter. The methods are `getAvailableDimensions` and `getColumnCapabilities`.

Reply:
Thanks. I changed.